app/src/main/java/com/example/composesigninsample/MainActivity.kt

```diff
@@ -109,22 +159,26 @@ class SignInViewModel @Inject constructor(
     }
 }
 
-/**
- * Home Screen and ViewModel
- */
 @Composable
-fun HomeScreen(navigateHome: () -> Unit, homeViewModel: HomeViewModel = hiltViewModel()) {
-    if (homeViewModel.loggedInState) {
-        Column {
-            Text(style = Typography.h1, text = "You are on the home screen!")
+fun HomeScreen(navigateToSignIn: () -> Unit, signOut: () -> Unit, isSignedIn: Boolean) {
+    if (isSignedIn) {
+        Column(horizontalAlignment = Alignment.CenterHorizontally, modifier = Modifier.padding(16.dp)) {
+            Text(style = Typography.h1, text = "You are on the home screen!", textAlign = TextAlign.Center)
+            Spacer(modifier = Modifier.height(8.dp))
             Text(style = Typography.body1, text = "Only for super secret signed in users only!")
-            Button(onClick = homeViewModel::signOut) {
+            Spacer(modifier = Modifier.height(16.dp))
+            Button(onClick = signOut) {
                 Text(text = "Sign Out")
             }
         }
-    } else {
-        LaunchedEffect(Unit) {
-            navigateHome()
+    }
+
+    DisposableEffect(isSignedIn) {
+        if (!isSignedIn) {
+            navigateToSignIn()
+        }
+
+        onDispose {
         }
     }
```

Comment:
Same as the above. I like that I can almost have a section of "effects" vs just having if statesments littered throughout.

Reply:
It's more accurate to say that if your effects aren't defined like this then you'll just have bugs

Reply:
Bugs you say... I wonder what kind would stem from this?

Reply:
Mostly that the effect shouldn't be hidden behind a conditional.

That's part of why your navigation didn't work previously.